fork-off/src/main.rs

```diff
@@ -12,168 +19,336 @@ pub struct Config {
     #[clap(long, default_value = "http://127.0.0.1:9933")]
     pub http_rpc_endpoint: String,
 
-    /// path to write the initial chainspec of the fork
-    /// as generated with the `bootstrap-chain` command
-    #[clap(long, default_value = "../docker/data/chainspec.json")]
-    pub fork_spec_path: String,
+    /// path of the initial chainspec (generated with the `bootstrap-chain` command)
+    #[clap(long, default_value = "./initial_chainspec.json")]
+    pub initial_spec_path: String,
+
+    /// where to write the snapshot of the state
+    #[clap(long, default_value = "./snapshot.json")]
+    pub snapshot_path: String,
+
+    /// where to write the forked genesis chainspec
+    #[clap(long, default_value = "./chainspec_from_snapshot.json")]
+    pub combined_spec_path: String,
 
     /// where to write the forked genesis chainspec
-    #[clap(long, default_value = "../docker/data/chainspec.fork.json")]
-    pub write_to_path: String,
+    #[clap(long)]
+    pub use_snapshot_file: bool,
+
+    /// how many parallel processes to download values -- note that large values might result in bans because
+    /// of rate-limiting mechanisms
+    #[clap(long, default_value_t = 5)]
+    pub num_workers: u32,
 
     /// which modules to set in forked spec
     #[clap(
         long,
         multiple_occurrences = true,
         takes_value = true,
         value_delimiter = ',',
-        default_value = "Aura,Aleph"
+        default_value = "Aura,Aleph,Sudo,Staking,Session,Elections"
     )]
-    pub prefixes: Vec<String>,
+    pub pallets: Vec<String>,
 }
 
-#[tokio::main]
-async fn main() -> anyhow::Result<()> {
-    let Config {
-        http_rpc_endpoint,
-        fork_spec_path,
-        write_to_path,
-        prefixes,
-    } = Config::parse();
+const KEYS_BATCH_SIZE: u32 = 1000;
 
-    env_logger::init();
+type BlockHash = String;
+type StorageKey = String;
+type StorageValue = String;
+type Storage = HashMap<StorageKey, StorageValue>;
 
-    info!(target: "fork",
-        "Running with config: \n\thttp_rpc_endpoint {}\n \tfork_spec_path: {}\n \twrite_to_path: {}\n \tprefixes: {:?}",
-        &http_rpc_endpoint, &fork_spec_path, &write_to_path, &prefixes
-    );
+async fn make_request_and_parse_result<T: serde::de::DeserializeOwned>(
+    client: &Client,
+    endpoint: &str,
+    body: Value,
+) -> T {
+    let mut response: Value = client
+        .post(endpoint)
+        .json(&body)
+        .send()
+        .await
+        .expect("Storage request has failed")
+        .json()
+        .await
+        .expect("Could not deserialize response as JSON");
+    let result = response["result"].take();
+    serde_json::from_value(result).expect("Incompatible type of the result")
+}
 
-    let mut fork_spec: Value = serde_json::from_str(
-        &fs::read_to_string(&fork_spec_path).expect("Could not read chainspec file"),
-    )?;
+fn construct_json_body(method_name: &str, params: Value) -> Value {
```

Comment:
this function is already in the aleph-client/rpc.rs file but is not yet made public. Let's use it here

Reply:
Ok, I see, but this is controversial as long as we are still planning to break the repo into smaller pieces. 

Also by depending upon `aleph_client` I automatically start depending on lots of substrate crates, which I wanted to avoid.

Reply:
Ok, Maybe as an improvement we can introduction features to aleph-client to not depend on substrate when we dont need it (not in this pr tho).